File: software/examples/rust/example_threshold.rs
use std::{io, error::Error};
use std::thread;
use tinkerforge::{ip_connection::IpConnection, 
                  industrial_dual_analog_in_bricklet::*};


const HOST: &str = "localhost";
const PORT: u16 = 4223;
const UID: &str = "XYZ"; // Change XYZ to the UID of your Industrial Dual Analog In Bricklet.

fn main() -> Result<(), Box<dyn Error>> {
    let ipcon = IpConnection::new(); // Create IP connection.
    let idai = IndustrialDualAnalogInBricklet::new(UID, &ipcon); // Create device object.

    ipcon.connect((HOST, PORT)).recv()??; // Connect to brickd.
    // Don't use device before ipcon is connected.

		// Get threshold receivers with a debounce time of 10 seconds (10000ms).
		idai.set_debounce_period(10000);

     let voltage_reached_receiver = idai.get_voltage_reached_callback_receiver();

        // Spawn thread to handle received callback messages. 
        // This thread ends when the `idai` object
        // is dropped, so there is no need for manual cleanup.
        thread::spawn(move || {
            for voltage_reached in voltage_reached_receiver {           
                		println!("Channel: {}", voltage_reached.channel);
		println!("Voltage: {} V", voltage_reached.voltage as f32 /1000.0);
		println!();
            }
        });

		// Configure threshold for voltage (channel 1) "greater than 10 V".
		idai.set_voltage_callback_threshold(1, '>', 10*1000, 0);

    println!("Press enter to exit.");
    let mut _input = String::new();
    io::stdin().read_line(&mut _input)?;
    ipcon.disconnect();
    Ok(())
}

File: software/examples/go/example_threshold.go
package main

import (
	"fmt"
	"github.com/Tinkerforge/go-api-bindings/industrial_dual_analog_in_bricklet"
	"github.com/Tinkerforge/go-api-bindings/ipconnection"
)

const ADDR string = "localhost:4223"
const UID string = "XYZ" // Change XYZ to the UID of your Industrial Dual Analog In Bricklet.

func main() {
	ipcon := ipconnection.New()
	defer ipcon.Close()
	idai, _ := industrial_dual_analog_in_bricklet.New(UID, &ipcon) // Create device object.

	ipcon.Connect(ADDR) // Connect to brickd.
	defer ipcon.Disconnect()
	// Don't use device before ipcon is connected.

	// Get threshold receivers with a debounce time of 10 seconds (10000ms).
	idai.SetDebouncePeriod(10000)

	idai.RegisterVoltageReachedCallback(func(channel uint8, voltage int32) {
		fmt.Printf("Channel: %d\n", channel)
		fmt.Printf("Voltage: %f V\n", float64(voltage)/1000.0)
		fmt.Println()
	})

	// Configure threshold for voltage (channel 1) "greater than 10 V".
	idai.SetVoltageCallbackThreshold(1, '>', 10*1000, 0)

	fmt.Print("Press enter to exit.")
	fmt.Scanln()

}
